src/gcp-residency.ts

```diff
@@ -0,0 +1,104 @@
+/**
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {execSync} from 'child_process';
+import {readFileSync} from 'fs';
+
+/**
+ * Known paths unique to Google Compute Engine Linux instances
+ */
+export const GCE_LINUX_BIOS_PATHS = {
+  BIOS_DATE: '/sys/class/dmi/id/bios_date',
+  BIOS_VENDOR: '/sys/class/dmi/id/bios_vendor',
+};
+
+/**
+ * Determines if the process is running on a Cloud Functions instance.
+ *
+ * Uses the {@link https://cloud.google.com/functions/docs/env-var Cloud Functions environment variables}.
+ *
+ * @returns {boolean} `true` if the process is running on Cloud Functions, `false` otherwise.
+ */
+export function isGoogleCloudFunction(): boolean {
+  /**
+   * `K_SERVICE` and `FUNCTION_NAME` are variables unique to Cloud Functions environments:
+   * - `FUNCTION_NAME` in {@link https://cloud.google.com/functions/docs/env-var Python 3.7 and Go 1.11}.
+   * - `K_SERVICE` in {@link https://cloud.google.com/functions/docs/env-var Newer runtimes}.
+   */
+  const isGFEnvironment = process.env.K_SERVICE || process.env.FUNCTION_NAME;
+
+  return !!isGFEnvironment;
+}
+
+/**
+ * Determines if the process is running on a Linux Google Compute Engine instance.
+ *
+ * @returns {boolean} `true` if the process is running on Linux Google Compute Engine, `false` otherwise.
+ */
+export function isGoogleComputeEngineLinux(): boolean {
+  if (process.platform !== 'linux') return false;
+
+  try {
+    // ensure this file exist
+    readFileSync(GCE_LINUX_BIOS_PATHS.BIOS_DATE, 'utf8');
+
+    // ensure this file exist and matches
+    const biosVendor = readFileSync(GCE_LINUX_BIOS_PATHS.BIOS_VENDOR, 'utf8');
+
+    return /Google/.test(biosVendor);
+  } catch {
+    return false;
+  }
+}
+
+/**
+ * Determines if the process is running on a Windows Google Compute Engine instance.
+ *
+ * @returns {boolean} `true` if the process is running on Windows GCE, `false` otherwise.
+ */
+export function isGoogleComputeEngineWindows(): boolean {
+  if (process.platform !== 'win32') return false;
```

Comment:
Given how expensive child processes are in node, it might be a nice thing to provide consumers a way to avoid this check via a env variable here, e.g:

```suggestion
  if (process.platform !== 'win32' || process.env.NO_GCE_CHECK) return false;
```

Reply:
Also, why the no-process suggestion was not chosen for Win as doc suggests?

Reply:
> Given how expensive child processes are in node, it might be a nice thing to provide consumers a way to avoid this check via a env variable here

I agree - I'll add an environment variable check here

> Also, why the no-process suggestion was not chosen for Win as doc suggests?

The recommended `ffi` project has been [long abandoned](https://github.com/node-ffi/node-ffi/issues/511). It's unfortunate as I haven't seen any other viable ffi or windows registry modules. As an additional constraint, we have to be careful with our dependencies for this library as dependency size and web compatibility are concerns for our JS customers.

I'll add this as a comment for discussing in the doc. In the meantime, I'll add a `do not merge` label to this PR until the discussion is resolved.

Reply:
One option is to exclude Windows for now and proceed.